pep-0544.txt

```diff
@@ -507,24 +507,29 @@ non-protocol generic types::
 ``Protocol[T, S, ...]`` is allowed as a shorthand for
 ``Protocol, Generic[T, S, ...]``.
 
-Declaring variance is not necessary for protocol classes, since it can be
-inferred from a protocol definition. Examples::
+User-defined generic protocols support explicitly declared variance.
+Type checkers will warn if an inferred variance is different from
+the declared variance. Examples::
 
-  class Box(Protocol[T]):
-      def content(self) -> T:
+  T = TypeVar('T')
+  T_co = TypeVar('T_co', covariant=True)
+  T_contra = TypeVar('T_contra', contravariant=True)
+
+  class Box(Protocol[T_co]):
+      def content(self) -> T_co:
```

Comment:
I think everything should be OK in this case (at least I can't invent any problematic example). Invariant subclass of a covariant class, and invariant subclass of a contravriant class are both OK (for nominal classes and for protocols).

At the same time going from covariant to contravariant or vice-versa is not safe, as well as covariant and contravariant subclasses of an invariant class are not safe.

Actually @pkch proposed a documentation PR about this some time ago https://github.com/python/mypy/pull/3179, but we decided to close it since mypy does not yet support these checks (variance overrides) for _nominal_ classes. Instead it was proposed to put this discussion in PEP 483. I suppose @pkch would like to make a PR to the peps repo.